File: Permutations.java
/**
 * LeetCode 46 Permutations
 */
class Solution {
    private void backtrack(int maxLength, int cur,List<Integer> nums, List<List<Integer>> result) {
        if (maxLength == cur) {
            result.add(new ArrayList<>(nums));
            return ;
        }
        for (int i = cur; i < maxLength; i++) {
            Collections.swap(nums,cur,i);
            backtrack(maxLength, cur + 1, nums, result);
            Collections.swap(nums,cur,i);
        }
        return ;
    }

    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> container = new ArrayList<>();
        if (nums == null || nums.length == 0) {
            return result;
        }

        for (int num : nums) {
            container.add(num);
        }
        backtrack(nums.length,0,container,result);
        return result;
    }
}

File: Permutations.go
func permute(nums []int) [][]int {
	res := make([][]int,0)
	if (nums == nil || len(nums) == 0) {
		return res
	}

	var backtrack func ([] int,int)

	backtrack = func(nums[] int,curIdx int) {
		if (curIdx == len(nums) - 1) {
			var temp = make([]int,len(nums))
			copy(temp,nums)
			res = append(res,temp)
			return
		}

		for i := curIdx;i < len(nums);i++ {
			nums[i],nums[curIdx] = nums[curIdx],nums[i]
			backtrack(nums,curIdx + 1)
			nums[i],nums[curIdx] = nums[curIdx],nums[i]
		}
	}

	backtrack(nums,0)
	return res
}


